--- a/drivers/hid/hid-nintendo.c	2021-12-08 00:16:59.019141089 +0100
+++ b/drivers/hid/hid-nintendo.c	2021-12-08 00:17:35.979378013 +0100
@@ -402,6 +402,15 @@ static const u16 JC_RUMBLE_DFLT_HIGH_FRE
 static const u16 JC_RUMBLE_PERIOD_MS = 50;
 static const unsigned short JC_RUMBLE_ZERO_AMP_PKT_CNT = 5;
 
+/* Used for player LEDs as found on game controllers from e.g. Nintendo, Sony. */
+#ifndef LED_FUNCTION_PLAYER1 // from include/dt-bindings/leds/common.h
+#define LED_FUNCTION_PLAYER1 "player-1"
+#define LED_FUNCTION_PLAYER2 "player-2"
+#define LED_FUNCTION_PLAYER3 "player-3"
+#define LED_FUNCTION_PLAYER4 "player-4"
+#define LED_FUNCTION_PLAYER5 "player-5"
+#endif
+
 static const char * const joycon_player_led_names[] = {
 	LED_FUNCTION_PLAYER1,
 	LED_FUNCTION_PLAYER2,
